core/services/gateway/handlers/webapicapabilities/handler.go

```diff
@@ -176,12 +215,117 @@ func (h *handler) handleWebAPITriggerMessage(ctx context.Context, msg *api.Messa
+		h.lggr.Errorw("error decoding payload", "err", err, "payload", string(body.Payload))
+		// callbackCh <- handlers.UserCallbackPayload{Msg: msg, ErrCode: api.UserMessageParseError, ErrMsg: fmt.Sprintf("error decoding payload %s", err.Error())}
+		// close(callbackCh)
+		return err
+	}
+	triggersConfig := make(map[string]webapicap.TriggerConfig)
+	for triggerID, configPbString := range payload {
+		pbBytes, err := base64.StdEncoding.DecodeString(configPbString)
+		if err != nil {
+			h.lggr.Errorw("error decoding pb bytes", "err", err)
+			return err
+		}
+		pb := &pb.Map{}
+		proto.Unmarshal(pbBytes, pb)
+		vmap, err := values.FromMapValueProto(pb)
+		if err != nil {
+			h.lggr.Errorw("error FromMapValueProto", "err", err)
+			return err
+		}
+		if vmap == nil {
+			h.lggr.Errorw("error FromMapValueProto nil vmap")
+			return err
+		}
+		reqConfig, err := h.ValidateConfig(vmap)
+		if err != nil {
+			h.lggr.Errorw("error validating config", "err", err)
+			return err
+		}
+		triggersConfig[triggerID] = *reqConfig
+	}
+	h.triggersConfig.triggersConfigMap[donID] = TriggersConfig{lastUpdatedAt: time.Now(), triggersConfig: triggersConfig}
+
+	// h.updateTriggerConsensus()
+	return nil
+}
+
+func (h *handler) updateTriggerConsensus() {
+	// calculate the mode of the trigger configs to determine consensus
+
+	// 1. Make list of node configs that haven't expired
+	triggers := h.triggersConfig.triggersConfigMap
+
+	// 2. Make list of hashes of each node's config.
+	// 3. Count the number of times each hash appears.
+
+	triggersHashmap := make(map[string]int)
+	triggersHashmapByNodeId := make(map[string]string)
+
+	for index, triggerConfig := range triggers {
+		s := fmt.Sprintf("%v", triggerConfig.triggersConfig)
+		h := sha1.New()
+		h.Write([]byte(s))
+		hash := hex.EncodeToString(h.Sum(nil))
+		triggersHashmap[hash]++
+		// the node that has this hash
+		triggersHashmapByNodeId[hash] = index
+	}
+
+	// 4. Find the hash that appears the most.
+	// https://stackoverflow.com/questions/18695346/how-can-i-sort-a-mapstringint-by-its-values
+
+	type kv struct {
+		Hash  string
+		Value int
+	}
+	var ss []kv
+	for k, v := range triggersHashmap {
+		ss = append(ss, kv{k, v})
+	}
+
+	sort.Slice(ss, func(i, j int) bool {
+		return ss[i].Value > ss[j].Value
+	})
+	// 5. Check if the hash that appears the most appears more than F times.
+	if ss[0].Value < int(h.config.F)+1 {
+		// 6. If it doesn't, do nothing
+		return
+	}
+	// 7. If it does, update the consensus config to that config.
+	NodeId := triggersHashmapByNodeId[ss[0].Hash]
+	h.consensusConfig = h.triggersConfig.triggersConfigMap[NodeId]
```

Comment:
I'm concerned that coming to consensus on an exact match of the whole blob of configs at once, which is a constantly changing map, may end up being flaky at higher trigger / node counts.

I believe the design intention was to aggregate them at more granular. Quoting the design doc:
"Each Gateway aggregates that metadata per workflow ID using MODE"

Reply:
I had talked to Bolek and he had that concern which is why he wrote that but thought my attempted simplification could be workable for a V1.

Reply:
I think this could work for V1.

Though counter argument: doing it by trigger probably isn't that much more complicated. Just iterate through them. Most complex part is that it might need a notion of the gateway being the source of truth of what to check. So when create trigger requests come through the Gateway, the Gateway stores its own list. And prunes it own list.

Reply:
I'm checking with Bolek about the use of Workflow Id, as I'm currently storing by trigger Id.